connect/runtime/src/main/java/org/apache/kafka/connect/util/LoggingContext.java

```diff
@@ -0,0 +1,229 @@
+     * {@link Parameters#CONNECTOR_TASK task number} to the same values in the supplied {@link ConnectorTaskId}.
+     *
+     * @param id the connector task ID; may be null
+     * @return the resulting LoggingContext that is already {@link LoggingContext#start() started}; never null
+     */
+    public static LoggingContext forTask(ConnectorTaskId id) {
+        Map<String, String> context = new HashMap<>();
+        context.put(Parameters.CONNECTOR_NAME, id.connector());
+        context.put(Parameters.CONNECTOR_TASK, Integer.toString(id.task()));
+        context.put(Parameters.CONNECTOR_SCOPE, Scope.TASK.value());
+        return new LoggingContext(context).start();
+    }
+
+    /**
+     * Modify the current {@link MDC} logging context to set the {@link Parameters#CONNECTOR_TYPE connector type} to a shortened name
+     * for the supplied class and the {@link Parameters#CONNECTOR_NAME connector name} to the supplied name.
+     *
+     * @param connectorClassName the fully qualified or shortened connector class name; may be null
+     * @return the resulting LoggingContext that is already {@link LoggingContext#start() started}; never null
+     */
+    public static LoggingContext forConnector(String connectorClassName, String connectorName) {
+        Map<String, String> context = new HashMap<>();
+        context.put(Parameters.CONNECTOR_TYPE, shortNameFor(connectorClassName));
+        context.put(Parameters.CONNECTOR_NAME, connectorName);
+        context.put(Parameters.CONNECTOR_SCOPE, Scope.WORKER.value());
+        return new LoggingContext(context).start();
+    }
+
+    /**
+     * Modify the current {@link MDC} logging context to set the  {@link Parameters#CONNECTOR_NAME connector name},
+     * {@link Parameters#CONNECTOR_TASK task number} and to the same values in the supplied {@link ConnectorTaskId}, and to set
+     * the scope to {@link Scope#OFFSETS}.
+     *
+     * @param id the connector task ID; may be null
+     * @return the resulting LoggingContext that is already {@link LoggingContext#start() started}; never null
+     */
+    public static LoggingContext forOffsets(ConnectorTaskId id) {
+        Map<String, String> context = new HashMap<>();
+        context.put(Parameters.CONNECTOR_NAME, id.connector());
+        context.put(Parameters.CONNECTOR_TASK, Integer.toString(id.task()));
+        context.put(Parameters.CONNECTOR_SCOPE, Scope.OFFSETS.value());
+        return new LoggingContext(context).start();
+    }
+
+    protected static String shortNameFor(String className) {
+        return className.replaceAll("^(.*)[.]", "") // greedy wildcard
+                        .replaceAll("^(.*)[$]", "") // greedy wildcard
+                        .replaceAll("Connector", "")
+                        .replaceAll("Task", "");
+    }
+
+    private static final String NULL_VALUE = null;
+
+    final Map<String, String> previous = new HashMap<>();
```

Comment:
Because when you use `getCopyOfContextMap` you get all of the properties and you still have to unset all of them and then put all of what's in the copy. However, that doesn't work as well if you're incrementally adding/removing a subset of the contexts, and it doesn't work if there are other components (e.g., connectors) that are adding contexts after we have.

The current approach is a bit more complicated, but it should be a lot easier to be completely independent of all other modifications of the MDC context map.